include/boost/math/tools/roots.hpp

```diff
@@ -647,7 +652,9 @@ namespace detail {
          }
          else if (result > max)
          {
-            T diff = ((fabs(max) < 1) && (fabs(result) > 1) && (tools::max_value<T>() / fabs(result) < fabs(max))) ? T(1000) : T(result / max);
+            const T amax = fabs(max);
+            volatile const T aresult = fabs(result); // volatile: force compiler to honor data-dependency in chained bool exprs below
```

Comment:
You should really only need `const volatile` in a shared memory environment. Do you still get overflow errors with just `const`?

Reply:
Yes, I tried all the variations I could think of.  What also worked was predclaring the `amax` and `aresult` and switching around `tools::max_value<T>() / fabs(result) < fabs(max)` to `tools::max_value<T>() / amax < aresult`, but I that was only suppressing it in this case for a specific relationship between `result` and `max`, so the `volatile` in my mind makes it a little more generic